### PR TITLE
Upload blacklists

### PR DESCRIPTION
**Pull request by shanu17**
[SYSDEV-876] Allowing users to upload blacklist files

[SYSDEV-876]: https://ulstracker.atlassian.net/browse/SYSDEV-876

### CODE REVIEW

File: BetterPasswordHandler.inc.php [hunk 1]
```diff
@@ -42,4 +42,48 @@ function signIn($args, $request) {
 		));
 		$templateMgr->display('frontend/pages/userLogin.tpl');
 	}
+        
+        /**
+        * Store the uploaded blacklists files
+        * @return boolean JSONMessage
```
Comment (ctgraham) [inline BetterPasswordHandler.inc.php:48]: Document your `@param`s here.

File: BetterPasswordHandler.inc.php [hunk 2]
```diff
@@ -42,4 +42,48 @@ function signIn($args, $request) {
...
+        */
+        function uploadBlacklists($args, $request) {
+            import('lib.pkp.classes.file.PrivateFileManager');
+            $privateFileManager = new PrivateFileManager();
+            
+            $uploadedFile = $_FILES['uploadedFile'];
+            $destFilePath = $privateFileManager->getBasePath() . DIRECTORY_SEPARATOR . 'betterPassword' . DIRECTORY_SEPARATOR . 'blacklists' . DIRECTORY_SEPARATOR . $uploadedFile['name'];
```
Comment (ctgraham) [inline BetterPasswordHandler.inc.php:55]: Doesn't the `$_FILES['uploadedFile']['name']` value come directly from the end user?  How can we trust this value?  What if a malicious user submitted "../../../../../../../etc/passwd"?

File: BetterPasswordPlugin.inc.php [hunk 3]
```diff
@@ -56,6 +56,7 @@ function register($category, $path, $mainContextId = null) {
 				HookRegistry::register('LoadHandler', array($this, 'callbackLoadHandler'));
 			}
 			HookRegistry::register('userdao::getAdditionalFieldNames', array(&$this, 'addUserSettings'));
+                        HookRegistry::register('LoadComponentHandler', array($this, 'callbackLoadHandler'));
```
Comment (ctgraham) [inline BetterPasswordPlugin.inc.php:59]: There are spaces rather than tabs here and elsewhere.

File: BetterPasswordSettingsForm.inc.php [hunk 4]
```diff
@@ -84,7 +84,9 @@ function readInputData() {
 	 * @copydoc Form::fetch()
 	 */
 	function fetch($request, $template = NULL, $display = false) {
+		$router = $request->getRouter();
 		AppLocale::requireComponents(LOCALE_COMPONENT_PKP_ADMIN);
+                $plugin = PluginRegistry::getPlugin('generic', 'betterpasswordplugin');
```
Comment (ctgraham) [inline BetterPasswordSettingsForm.inc.php:89]: The plugin is available here from `$this->_plugin`.

File: templates/settingsForm.tpl [hunk 5]
```diff
@@ -9,8 +9,16 @@
  *}
 <script>
 	$(function() {ldelim}
-		// Attach the form handler.
-		$('#betterPasswordSettingsForm').pkpHandler('$.pkp.controllers.form.AjaxFormHandler');
```
Comment (ctgraham) [inline templates/settingsForm.tpl:13]: Is the delete of this `pkpHandler` intentional?

File: BetterPasswordHandler.inc.php [hunk 6]
```diff
@@ -42,4 +42,48 @@ function signIn($args, $request) {
...
+                    $prevBlacklist[$destFilePath] = sha1_file($destFilePath);
+                    $plugin->updateSetting(CONTEXT_SITE, 'betterPasswordUserBlacklistFiles', $prevBlacklist);
+                }
+            }
+            return new JSONMessage(true);
+        }
+
+        /**
+        * Store the uploaded blacklists files
+        * @return boolean JSONMessage
```
Comment (ctgraham) [inline BetterPasswordHandler.inc.php:72]: Document `@param`s here.

File: templates/settingsForm.tpl (hunk 5 above)
Comment (shanu17) [reply templates/settingsForm.tpl:13]: The form functions as expected without this Ajax Handler. The only handler required is the file upload handler.

Comment (ctgraham) [reply templates/settingsForm.tpl:13]: Can you review and share the function of the AjaxFormHander and document why it is not needed here?

Comment (shanu17) [reply templates/settingsForm.tpl:13]: https://github.com/pkp/pkp-lib/blob/master/js/controllers/form/FileUploadFormHandler.js#L19 
The file upload handler extends the Ajax Form Handler

File: BetterPasswordHandler.inc.php [hunk 7]
```diff
@@ -42,4 +42,49 @@ function signIn($args, $request) {
 		));
 		$templateMgr->display('frontend/pages/userLogin.tpl');
 	}
+        
+	/**
+	 * Store the uploaded blacklists files
+	 * @param $args array Arguments array expecting user uploaded file properties
+	 * @param $request PKPRequest Request object.
+	 * @return boolean JSONMessage
```
Comment (ctgraham) [inline BetterPasswordHandler.inc.php:50]: This returns `JSONMessage`, not `boolean`.

File: BetterPasswordHandler.inc.php [hunk 8]
```diff
@@ -42,4 +42,49 @@ function signIn($args, $request) {
...
+			$plugin->updateSetting(CONTEXT_SITE, 'betterPasswordUserBlacklistFiles', $prevBlacklist);
+		}
+		return new JSONMessage(true);
+	}
+
+	/**
+	 * Delete the user uploaded blacklists files
+	 * @param $args array Arguments array expecting user uploaded file hash
+	 * @param $request PKPRequest Request object.
+	 * @return boolean JSONMessage
```
Comment (ctgraham) [inline BetterPasswordHandler.inc.php:72]: This also returns `JSONMessage` rather than `boolean`.

File: BetterPasswordPlugin.inc.php [hunk 9]
```diff
@@ -214,9 +215,16 @@ function checkPassword($hookName, $args) {
 	 * @return array filename strings
 	 */
 	function getBlacklists() {
-		return array(
-			$this->getPluginPath() . DIRECTORY_SEPARATOR . 'badPasswords' . DIRECTORY_SEPARATOR . 'badPasswords.txt',
-		);
+			import('lib.pkp.classes.file.PrivateFileManager');
```
Comment (ctgraham) [inline BetterPasswordPlugin.inc.php:218]: Some extra indentation here.

File: BetterPasswordPlugin.inc.php [hunk 10]
```diff
@@ -287,7 +295,17 @@ public function callbackLoadHandler($hookName, $args) {
 					$user->setData($this->getName()."::badPasswordTime", 0);
 					$userDao->updateObject($user);
 			}
-		}
+		} elseif($hookName === "LoadComponentHandler" && $args[1] === "uploadBlacklists") {
+                        define('HANDLER_CLASS', 'BetterPasswordComponentHandler');
```
Comment (ctgraham) [inline BetterPasswordPlugin.inc.php:299]: Return of the spaces.

File: BetterPasswordSettingsForm.inc.php [hunk 11]
```diff
@@ -84,7 +84,9 @@ function readInputData() {
 	 * @copydoc Form::fetch()
 	 */
 	function fetch($request, $template = NULL, $display = false) {
+		$router = $request->getRouter();
 		AppLocale::requireComponents(LOCALE_COMPONENT_PKP_ADMIN);
+                $plugin = $this->_plugin;
```
Comment (ctgraham) [inline BetterPasswordSettingsForm.inc.php:89]: More spaces here.

File: locale/en_US/locale.xml [hunk 12]
```diff
@@ -22,13 +22,18 @@
 	<message key="plugins.generic.betterPassword.validation.betterPasswordCheckSpecial">Your password requires an non-alphanumeric character.</message>
 	<message key="plugins.generic.betterPassword.validation.betterPasswordCheckBlacklist">Your password was found in a blacklist of known bad passwords.</message>
 	<message key="plugins.generic.betterPassword.validation.betterPasswordLocked">Your account has been temporarily locked due to bad password attempts.</message>
+        <message key="plugins.generic.betterPassword.validation.betterPasswordUnexpectedError">Your password could not be changed at this time.  Please try again later.</message>
```
Comment (ctgraham) [inline locale/en_US/locale.xml:25]: Spaces here.

File: templates/settingsForm.tpl [hunk 13]
```diff
@@ -22,7 +30,16 @@
 	{fbvFormArea id="betterPasswordSettingsFormOptions" title="admin.siteSettings"}
 		{fbvFormSection for="minPasswordLength"}
 			{fbvElement type="text" id="minPasswordLength" label="admin.settings.minPasswordLength" value="$minPasswordLength"}
-		{/fbvFormSection}
+                {/fbvFormSection}
```
Comment (ctgraham) [inline templates/settingsForm.tpl:33]: Spaces here.

File: templates/settingsForm.tpl [hunk 14]
```diff
@@ -22,7 +30,16 @@
...
+                {fbvFormSection title="plugins.generic.betterPassword.manager.settings.betterPasswordExistingBlacklist"}
+                        {foreach from=$betterPasswordBlacklistFiles key="betterPasswordFile" item="betterPasswordSettingValue"}
+                                {assign "temp" ["fileId" => $betterPasswordSettingValue]}
```
Comment (ctgraham) [inline templates/settingsForm.tpl:36]: This `$temp` variable does not appear to be used.

**Pull request merged**